### PR TITLE
RavenDB-8404,  RavenDB-9134

### PR DESCRIPTION
**Pull request by maximburyak**

### CODE REVIEW

File: src/Raven.Server/Documents/Patch/ScriptRunner.cs [hunk 1]
```diff
@@ -426,6 +429,16 @@ private static JsValue EndsWith(JsValue self, JsValue[] args)
                 return new JsValue(args[0].AsString().EndsWith(args[1].AsString(), StringComparison.OrdinalIgnoreCase));
             }
 
+            private JsValue Regex(JsValue self, JsValue[] args)
+            {
+                if (args.Length != 2 || args[0].IsString() == false || args[1].IsString() == false)
+                    throw new InvalidOperationException("startsWith(text, contained) must be called with two string paremters");
```
Comment (ayende) [inline src/Raven.Server/Documents/Patch/ScriptRunner.cs:435]: wrong method name

File: src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs [hunk 2]
```diff
@@ -95,9 +100,72 @@ public override void VisitCompoundWhereExpression(BinaryExpression @where)
 
         public override void VisitMethod(MethodExpression expr)
         {
+            // The regex usage here is just temporary, to get things working. Ideally, we want to implement the string functions in Jint
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs:103]: Bad comment?

File: src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs [hunk 3]
```diff
@@ -95,9 +100,72 @@ public override void VisitCompoundWhereExpression(BinaryExpression @where)
...
+            if (expr.Name.Value.Equals("startswith", StringComparison.OrdinalIgnoreCase) && expr.Arguments.Count == 2)
+            {
+                _sb.Append("startsWith(");
+                VisitExpression(expr.Arguments[0]);
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs:107]: Don't you need to escape the variable here, what if we have a quote in it?

File: src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs [hunk 4]
```diff
@@ -95,9 +100,72 @@ public override void VisitCompoundWhereExpression(BinaryExpression @where)
...
+                _sb.Append(",");
+                VisitExpression(expr.Arguments[1]);
+                _sb.Append(")");
+                return;
+            }
+            
+            if (expr.Name.Value.Equals("endswith", StringComparison.OrdinalIgnoreCase) && expr.Arguments.Count == 2)
+            {
+                _sb.Append("endsWith(");
+                VisitExpression(expr.Arguments[0]);
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs:117]: Same

File: src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs [hunk 5]
```diff
@@ -95,9 +100,72 @@ public override void VisitCompoundWhereExpression(BinaryExpression @where)
...
+                _sb.Append(",");
+                VisitExpression(expr.Arguments[1]);
+                _sb.Append(")");
+                return;
+            }
+
+            if (expr.Name.Value.Equals("regex", StringComparison.OrdinalIgnoreCase) && expr.Arguments.Count == 2)
+            {
+                _sb.Append("regex(");
+                VisitExpression(expr.Arguments[0]);
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs:127]: same

File: src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs [hunk 6]
```diff
@@ -95,9 +100,72 @@ public override void VisitCompoundWhereExpression(BinaryExpression @where)
...
+                _sb.Append("!== 'undefined')");
+                return;    
+            }
+            
+            if (UnsopportedQueryMethodsInJavascript.Any(x=>
+                x.Equals(expr.Name.Value, StringComparison.OrdinalIgnoreCase)))
+            {
+                throw new NotSupportedException($"'{expr.Name.Value}' query method is not supported by subscriptions");
+            }
+
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs:164]: So `regex('a','b','c')` would be sent here?

File: src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs (hunk 3 above)
Comment (maximburyak) [reply src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs:107]: Not sure what you mean, escaping treament is in VisitValue

**Pull request merged**